pkg/sensors/load.go

```diff
@@ -149,6 +153,25 @@ func (s *Sensor) Unload() error {
 	return nil
 }
 
+// Destroy will unload the hook and call DestroyHook, this hook is usually used
+// to clean up resources that were created during creation of the sensor.
+func (s *Sensor) Destroy() error {
+	err := s.Unload()
+	if err != nil {
+		return err
+	}
+
+	if s.DestroyHook != nil {
+		err = s.DestroyHook()
+		if err != nil {
+			return err
+		}
+	}
+	s.Destroyed = true
```

Comment:
So what happens if either Unload or Destroy fails? do we have a way to recover?
I'm wondering whether it makes sense to keep the s.Destroyed state or we if it's better to just remove the sensor altogether if the unload or the hook fails.

Reply:
we can keep the `s.Destroyed` state indeed I agree but what do you imply by removing the sensor altogether? What could we do more?

Reply:
we could `defer s.Destroyed = true` at the beginning basically. 

I'm not sure it would be very fair for situations where it just failed for casual reasons like Unload failed for "sensor not loaded".

What would be your idea? To run the destroy hook and affect the Destroyed state even if the Unload failed? And Destroyed state even if the hook failed?

Reply:
So let's say user delete tracing policy. so we are in `deleteTracingPolicy`.

If everything goes well, we will call `delete(h.collections, op.name)` and the user will have no way of accessing the sensor.

If `s.Destroy()` returns an error, then the collection will not be deleted and the tracingpolicy will remain the sensor manager in a defunc state. There is nothing the user can do at this point.

So my two quetions are:
* Why do we keep the s.Destroyed variable? What is a scenario that we will end up using the value? (we reading it in Load(), so can we ever call Load after we call Destroy()? This seems weird to me. If a sensor is "destroyed", why are we holding a reference to it?
* What happens if unload hook fails? Then, the user will not be able to load a new policy with the same name and they will also not be able to remove the old policy that is now defunct. It seems better to me to remove the collection even if the Destroy() function fails.

Thoughts?

Reply:
ah yes I understand, I didn't think of `delete(h.collections, op.name)` because this is what is done by the caller of this function!

> Why do we keep the s.Destroyed variable? What is a scenario that we will end up using the value? (we reading it in Load(), so can we ever call Load after we call Destroy()? This seems weird to me. If a sensor is "destroyed", why are we holding a reference to it?

Yes agree, the `destroyed` state is just to prevent stuff and should not be triggered in normal use.

> What happens if unload hook fails? Then, the user will not be able to load a new policy with the same name and they will also not be able to remove the old policy that is now defunct. It seems better to me to remove the collection even if the Destroy() function fails.

Sure! We could basically do something like this:

```diff
func (h *handler) deleteTracingPolicy(op *tracingPolicyDelete) error {
	col, exists := h.collections[op.name]
	if !exists {
		return fmt.Errorf("tracing policy %s does not exist", op.name)
	}
+	defer delete(h.collections, op.name)
	
	err := col.destroy()
	if err != nil {
		col.err = fmt.Errorf("failed to unload tracing policy: %w", err)
		return err
	}

	filterID := policyfilter.PolicyID(col.policyfilterID)
	err = h.pfState.DelPolicy(filterID)
	if err != nil {
		col.err = fmt.Errorf("failed to remove from policyfilter: %w", err)
		return err
	}

-	delete(h.collections, op.name)
	return nil
}
```

Only from the `s.Destroy()` point of view, what do you think of the Destroyed state, should it automatically set when you call that even if it fails somehow. I just noticed that in `s.Unload()` the only returned error is `unload of sensor %s failed: sensor not loaded`, the rest is just logged for the failure of the PreUnload and PostUnload hooks. So maybe we could do something like that:

```diff
// Destroy will unload the hook and call DestroyHook, this hook is usually used
// to clean up resources that were created during creation of the sensor.
func (s *Sensor) Destroy() error {
	err := s.Unload()
	if err != nil {
		return err // <- only error possible is sensor was not loaded, should destroy fail for that? maybe it's okay to destroy an unloaded sensor and thus we can move the `s.Destroyed = true` before this.
	}

	if s.DestroyHook != nil {
	        err = s.DestroyHook()
+               logger.GetLogger().WithError(err).WithField("sensor", s.Name).Warn("Destroy hook failed")
-		if err != nil {
-			return err
-		}
	}
	s.Destroyed = true

	return nil
}
```

What do you think?

Reply:
if we don't accept any error from `Destroy` like it always succeed as noted in the comment above we can do that then:

```diff
func (h *handler) deleteTracingPolicy(op *tracingPolicyDelete) error {
	col, exists := h.collections[op.name]
	if !exists {
		return fmt.Errorf("tracing policy %s does not exist", op.name)
	}
+	defer delete(h.collections, op.name)
	
	err := col.destroy()
-	if err != nil {
-		col.err = fmt.Errorf("failed to unload tracing policy: %w", err)
-		return err
-	}

	filterID := policyfilter.PolicyID(col.policyfilterID)
	err = h.pfState.DelPolicy(filterID)
	if err != nil {
		col.err = fmt.Errorf("failed to remove from policyfilter: %w", err)
		return err
	}

-	delete(h.collections, op.name)
	return nil
}
```

When you delete a policy, is it important to know that you could not unload it because it was never loaded in the first place? I guess no. The only thing is that we are making a lot of assumptions about the return case that we call and this can change in the future.

Reply:
Above sounds good to me. The only minor comment would be that we probably want something like:
```
if (err) {
     logger.GetLogger().WithError(err).WithField("sensor", s.Name).Warn("Destroy hook failed")
 }
```
 
 in case err = s.Destroy() hook is not nill

Reply:
> in case err = s.Destroy() hook is not nill
ah yes yes sure, I wrote too fast

Reply:
What's your stance on ignoring or not the error from `s.Unload` in `s.Destroy`? It feels like we should, does this sound good to you? It would result in something like this:

```diff
func (h *handler) deleteTracingPolicy(op *tracingPolicyDelete) error {
	col, exists := h.collections[op.name]
	if !exists {
		return fmt.Errorf("tracing policy %s does not exist", op.name)
	}
+	defer delete(h.collections, op.name)
	
	col.destroy() // <- doesn't return an error
-	if err != nil {
-		col.err = fmt.Errorf("failed to unload tracing policy: %w", err)
-		return err
-	}

	filterID := policyfilter.PolicyID(col.policyfilterID)
	err = h.pfState.DelPolicy(filterID)
	if err != nil {
		col.err = fmt.Errorf("failed to remove from policyfilter: %w", err)
		return err
	}

-	delete(h.collections, op.name)
	return nil
}
```

Reply:
I agree. Let's go with that, and we can always change it if the need arises. Thanks!

Reply:
See the simplification in the new push https://github.com/cilium/tetragon/pull/1562/commits/03f40739b741a087c992dc01cce861f7ff921ddb. :)